### ISSUE TITLE
Bidi support for lists

### ISSUE THREAD

**Issue by iMrDJAi** [ahangarha/markdown-it-bidi#25]
[Lists](https://spec.commonmark.org/0.30/#lists) (\<ol>, \<ul>) are [container block elements](https://spec.commonmark.org/0.30/#container-blocks), they contain [list items](https://spec.commonmark.org/0.30/#list-items) (\<li>). List items contain text. However, they are also container block elements that may contain nested lists.

Lists should always take a single direction, which means if the direction of the first list item is right to left, the other list items should all have a right to left direction (and the opposite is true). Otherwise, they will look ugly, messy, and will be hard to read.

Adding `dir="auto"` to all block elements doesn't work well with lists. See this example:

Markdown:

```md
RTL:

- تجربة
- test

1. تجربة
2. test

LTR:

- test
- تجربة

1. test
2. تجربة
```

Result:

![image](https://user-images.githubusercontent.com/42304709/159015339-7d36e355-f3c5-49a8-8e3e-aa9ccc0a2bd4.png)

```html
<p dir="auto">RTL:</p>
<ul dir="auto">
  <li dir="auto">تجربة</li>
  <li dir="auto">test</li>
</ul>
<ol dir="auto">
  <li dir="auto">تجربة</li>
  <li dir="auto">test</li>
</ol>
<p dir="auto">LTR:</p>
<ul dir="auto">
  <li dir="auto">test</li>
  <li dir="auto">تجربة</li>
</ul>
<ol dir="auto">
  <li dir="auto">test</li>
  <li dir="auto">تجربة</li>
</ol>
```

Notice how the list markers are completely broken. To fix that, we should remove the `dir` attribute from list items:

![image](https://user-images.githubusercontent.com/42304709/159015975-04cb91eb-8cac-4782-9fce-3928794f1034.png)

```html
<p dir="auto">RTL:</p>
<ul dir="auto">
  <li>تجربة</li>
  <li>test</li>
</ul>
<ol dir="auto">
  <li>تجربة</li>
  <li>test</li>
</ol>
<p dir="auto">LTR:</p>
<ul dir="auto">
  <li>test</li>
  <li>تجربة</li>
</ul>
<ol dir="auto">
  <li>test</li>
  <li>تجربة</li>
</ol>
```

Another issue with nested lists:

Markdown:

```md
RTL:

- تجربة
- - test
- - - test
- - - تجربة

LTR:

- test
- - تجربة
- - - تجربة
- - - test
```

![image](https://user-images.githubusercontent.com/42304709/159020560-6730c314-6135-46d2-8385-b697ca542ba4.png)

```html
<p dir="auto">RTL:</p>
<ul dir="auto">
  <li>تجربة</li>
  <li>
    <ul dir="auto">
      <li>test</li>
    </ul>
  </li>
  <li>
    <ul dir="auto">
      <li>
        <ul dir="auto">
          <li>test</li>
        </ul>
      </li>
    </ul>
  </li>
  <li>
    <ul dir="auto">
      <li>
        <ul dir="auto">
          <li>تجربة</li>
        </ul>
      </li>
    </ul>
  </li>
</ul>
<p dir="auto">LTR:</p>
<ul dir="auto">
  <li>test</li>
  <li>
    <ul dir="auto">
      <li>تجربة</li>
    </ul>
  </li>
  <li>
    <ul dir="auto">
      <li>
        <ul dir="auto">
          <li>تجربة</li>
        </ul>
      </li>
    </ul>
  </li>
  <li>
    <ul dir="auto">
      <li>
        <ul dir="auto">
          <li>test</li>
        </ul>
      </li>
    </ul>
  </li>
</ul>
```

Some list markers are aligned on the opposite side. This is a total mess! We can fix that by removing the `dir` attribute from nested lists:

![image](https://user-images.githubusercontent.com/42304709/159024048-42efffb1-8c03-4eb8-aa13-bdb599785944.png)

```html
<p dir="auto">RTL:</p>
<ul dir="auto">
  <li>تجربة</li>
  <li>
    <ul>
      <li>test</li>
    </ul>
  </li>
  <li>
    <ul>
      <li>
        <ul>
          <li>test</li>
        </ul>
      </li>
    </ul>
  </li>
  <li>
    <ul>
      <li>
        <ul>
          <li>تجربة</li>
        </ul>
      </li>
    </ul>
  </li>
</ul>
<p dir="auto">LTR:</p>
<ul dir="auto">
  <li>test</li>
  <li>
    <ul>
      <li>تجربة</li>
    </ul>
  </li>
  <li>
    <ul>
      <li>
        <ul>
          <li>تجربة</li>
        </ul>
      </li>
    </ul>
  </li>
  <li>
    <ul>
      <li>
        <ul>
          <li>test</li>
        </ul>
      </li>
    </ul>
  </li>
</ul>
```

**Conclusion:** We should not give `dir="auto"` to any child block-elements of lists. Only top-level lists should have the `dir` attribute set to `auto`.

I see no other solutions for lists, we should go ahead and implement this one.

**Comment by iMrDJAi** [comment]
@ahangarha CC

**Comment by ahangarha** [comment]
Thank you for your insightful investigation.

Since before also we had discussion regarding nested elements, I think we need to drop support for nested elements and only keep adding `dir="auto"` only to the root level elements. This includes blockquote.

Supporting nested elements should be postponed.

**Comment by iMrDJAi** [comment]
@ahangarha So we will end up implementing it anyway 😅

At this point, there is no reason to postpone nested elements support, we are already 1 month late for our schedule and we still haven't made real progress yet. The solutions are available for both blockquotes and tables and it's all about making a decision. 

I believe that we should take a step forward and start preparing to write an initial specification for bidi-markdown. An incomplete spec isn't something that I'm willing for. I'm kind of in a rush since I'm supposed to implement bidi support for other libraries.

I'm going to write a full implementation in a separate branch then I'll let you review it. If you're okay with it we can go ahead and adopt it.

**Comment by ahangarha** [comment]
It both sounds good and bad. Good in the sense that you have ideas for implementation and bad in the sense that there is no communication on it. But it is good to make progress.

I would prefer to go slow with one or two libraries and find issues. I believe we need to take some steps in improving standards. In this sense I think it is better to have limited projects involved so that any future breaking change would have lesser negative impacts.

I suggest to not speed up this fast. It might make our work more difficult in future.

**Comment by iMrDJAi** [comment]
> bad in the sense that there is no communication

@ahangarha Yeah we should have a discussion about it, the thing is I just wanted to give you enough time to take a look at my implementation to have a better idea about it.

> I would prefer to go slow with one or two libraries and find issues

Of course, by "wanting to implement bidi support for other libraries" I meant one specific library that I'm interested to work on ([comrak](https://github.com/kivikakk/comrak)). Otherwise, I can't think of any other libraries (except [marked](https://github.com/markedjs/marked)).
